### PR TITLE
Centralize operator manifest generation

### PR DESCRIPTION
**Pull request by damemi**
Describes a proposal for moving operator CRD manifests to the `openshift/api` repo

### CODE REVIEW

File: enhancements/centralized-manifest-openapi-generation.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+- [ ] Enhancement has been determined `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is defined
+
+## Summary
+This document proposes migrating operator CRD manifests (and, more importantly, the work associated with updating them)
+to be colocated with the Golang types that define them in the [openshift/api](https://github.com/openshift/api) repository.
+
+## Motivation
+We have many operators that define custom resources, but almost all of those resources are based on Golang types defined in
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: not just operators.  The openshift/api is made up of many custom resources.

File: enhancements/centralized-manifest-openapi-generation.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+`openshift/api`. When those Golang types change, or CRD requirements change such as with [Kubernetes 1.16's
+structural schemas](https://kubernetes.io/blog/2019/06/20/crd-structural-schema/), these changes require matching updates
+to each operator repo. This is primarily evident in changes to OpenAPIV3 validation schemas, where it is not uncommon
+to make small changes to API markers which then need to be individually carried to each operator where they are generated
+into new manifest changes. In addition, maintainers currently may have different approaches to generating their manifests
+(using different tools and scripts) which can lead to discrepancies and unexpected behavior without support.
+
+By moving the operator manifests to `openshift/api`, we will colocate these types with their consumers and enable
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: all CRD manifests

File: enhancements/centralized-manifest-openapi-generation.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+organization-wide changes to these manifests to be enabled at once, included with the corresponding changes to their types.
+We also will provide a single approach to maintaining the API/manifest relationship, specifically in regards to generating
+OpenAPIV3 validation schemas.
+
+### Goals
+
+- Operator and API maintainers can make changes to APIs and their corresponding CRD manifests simultaneously
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: ordered list please

File: enhancements/centralized-manifest-openapi-generation.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+- Unify organization schema generation under a single tool and approach for CRD schema generation
+
+### Non-goals
+
+- We will not be responsible for maintaining individual component APIs or manifests. Though we are providing a single method
+to organize their management, maintainers will still be responsible for changes related to only their operator.
+
+## Proposal
+
+First, operator manifests will be moved to subdirectories of `openshift/api` corresponding to the names of the operators.
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: give an example. This seems wrong.

Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: dont' focus on operators. We don't care if they are operators.  In fact, we don't even really want the operator ones. We want the config.openshift.io ones.

File: enhancements/centralized-manifest-openapi-generation.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Then, makefile targets will be added to `openshift/api` referencing [existing build machinery targets](https://github.com/openshift/library-go/blob/27b03247913b2f6a0abcea95cb9b6c896adc2531/alpha-build-machinery/make/targets/openshift/crd-schema-gen.mk) that enable OpenAPI schema generation.
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: you are not allowed to create an import cycle

File: enhancements/centralized-manifest-openapi-generation.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+With the combination of these targets and separate manifest directories, we will be able to configure variables which allow
+us to update specific components' manifests (`update-crd-schema-$(1)` for example, `make update-crd-schema-cluster-kube-apiserver-operator`)
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: just generate everything.

File: enhancements/centralized-manifest-openapi-generation.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+- Operator and API maintainers can make changes to APIs and their corresponding CRD manifests simultaneously
+- Manifest changes can be easily read from `openshift/api` to the operators that consume them
+- Unify organization schema generation under a single tool and approach for CRD schema generation
+
+### Non-goals
+
+- We will not be responsible for maintaining individual component APIs or manifests. Though we are providing a single method
+to organize their management, maintainers will still be responsible for changes related to only their operator.
+
+## Proposal
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: How about just generating by group and dropping the crd manifests into the same package?

Then for the cluster-config-operator case, we simply have a bash script that copies those out?

File: enhancements/centralized-manifest-openapi-generation.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	$(foreach p,$(crd_patches),$(call patch-crd,$(basename $(p)).yaml,$(p)))
+.PHONY: update-codegen-crds
+```
+
+Operators will then reference these manifests by vendoring their respective packages and either
+1. Symlinking their current manifest directories to the vendored path or
+2. Referencing the vendored path directly in their code
+
+A third option would be to also generate the relevant manifest bindata in `openshift/api`, in which case the consuming operators
+could simply import the package for their bindata and use it directly.
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: bindata doesn't work for our first consumer, cluster-config-operator.

File: enhancements/centralized-manifest-openapi-generation.md (hunk 7 above)
Comment (damemi) [reply enhancements/centralized-manifest-openapi-generation.md]: The `controller-tools` generator provides input/output directory options to drop the manifests into, it still generates by group but I don't see any reason to generate into a folder that isn't our intended final destination

File: enhancements/centralized-manifest-openapi-generation.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+to be colocated with the Golang types that define them in the [openshift/api](https://github.com/openshift/api) repository.
+
+## Motivation
+We have many operators that define custom resources, but almost all of those resources are based on Golang types defined in
+`openshift/api`. When those Golang types change, or CRD requirements change such as with [Kubernetes 1.16's
+structural schemas](https://kubernetes.io/blog/2019/06/20/crd-structural-schema/), these changes require matching updates
+to each operator repo. This is primarily evident in changes to OpenAPIV3 validation schemas, where it is not uncommon
+to make small changes to API markers which then need to be individually carried to each operator where they are generated
+into new manifest changes. In addition, maintainers currently may have different approaches to generating their manifests
+(using different tools and scripts) which can lead to discrepancies and unexpected behavior without support.
```
Comment (sttts) [inline enhancements/centralized-manifest-openapi-generation.md]: this last sentence is the main reason IMO.

**Comment by damemi** [comment]
@deads2k so this is *only* for the config resources, we don't want to migrate *every* CRD to be generated in openshift/api?

File: enhancements/centralized-manifest-openapi-generation.md (hunk 1 above)
Comment (damemi) [reply enhancements/centralized-manifest-openapi-generation.md]: yeah I wasn't implying that openshift/api is only operator resources, but that all the operator resources are in openshift/api. I'll clarify the wording

File: enhancements/centralized-manifest-openapi-generation.md (hunk 4 above)
Comment (deads2k) [reply enhancements/centralized-manifest-openapi-generation.md]: why not embed the yaml in the same package.  Same package lets owners be in the correct spot.

File: enhancements/centralized-manifest-openapi-generation.md (hunk 5 above)
Comment (deads2k) [reply enhancements/centralized-manifest-openapi-generation.md]: > Would we just copy over the makefile targets to run the tool in openshift/api then?

If necessary, yes.

File: enhancements/centralized-manifest-openapi-generation.md (hunk 4 above)
Comment (damemi) [reply enhancements/centralized-manifest-openapi-generation.md]: that's a better plan, updated

File: enhancements/centralized-manifest-openapi-generation.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+1. Operator and API maintainers can make changes to APIs and their corresponding CRD manifests simultaneously
+2. Manifest changes can be easily read from `openshift/api` to the operators that consume them
+3. Unify organization schema generation under a single tool and approach for CRD schema generation
+
+### Non-goals
+
+- We will not be responsible for maintaining individual component APIs or manifests. Though we are providing a single method
+to organize their management, maintainers will still be responsible for changes related to only their operator.
+
+## Proposal
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: please provide a directory example.  Fill in a directory with psuedo files and run `tree config/`.

File: enhancements/centralized-manifest-openapi-generation.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+CRD manifests from relevant repositories will be moved to the same package  directories in `openshift/api` as the types that define them.
+This provides the benefit of allowing these types to be imported with Go modules, as well as inheriting the existing OWNERS
+files for those packages.
+
+Then, makefile targets will be added to `openshift/api` based on the [existing build machinery targets](https://github.com/openshift/library-go/blob/27b03247913b2f6a0abcea95cb9b6c896adc2531/alpha-build-machinery/make/targets/openshift/crd-schema-gen.mk) that currently enables OpenAPI schema generation. This generator will then be run when any API changes require
+an update to CRD manifests, and manifests will be verified by CI, similar to how these changes are currently tested in many repos.
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: indicate that we will enable this on a per API group level and state your intent of which file will maintain the list of CRD generated groups or directories.  I don't actually know which.

File: enhancements/centralized-manifest-openapi-generation.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+Example, for config manifests:
+```
+config/
+├── install.go
+└── v1
+    ├── 0000_03_authorization-openshift_01_rolebindingrestriction.crd.yaml
+    ├── 0000_03_config-operator_01_operatorhub.crd.yaml
+    ├── 0000_03_config-operator_01_proxy.crd.yaml
+    ├── 0000_03_quota-openshift_01_clusterresourcequota.crd.yaml
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: this doesn't belong here.

File: enhancements/centralized-manifest-openapi-generation.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+CRD manifests from relevant repositories will be moved to the same package  directories in `openshift/api` as the types that define them.
+This provides the benefit of allowing these types to be imported with Go modules, as well as inheriting the existing OWNERS
+files for those packages.
+
+Example, for config manifests:
+```
+config/
+├── install.go
+└── v1
+    ├── 0000_03_authorization-openshift_01_rolebindingrestriction.crd.yaml
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: not in this folder

File: enhancements/centralized-manifest-openapi-generation.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    ├── 0000_03_config-operator_01_operatorhub.crd.yaml
+    ├── 0000_03_config-operator_01_proxy.crd.yaml
+    ├── 0000_03_quota-openshift_01_clusterresourcequota.crd.yaml
+    ├── 0000_03_security-openshift_01_scc.crd.yaml
+    ├── 0000_10_config-operator_01_apiserver.crd.yaml
+    ├── 0000_10_config-operator_01_authentication.crd.yaml
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: the generator will file them with these number prefixes?

File: enhancements/centralized-manifest-openapi-generation.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+    ├── 0000_10_config-operator_01_featuregate.crd.yaml
+    ├── 0000_10_config-operator_01_imagecontentsourcepolicy.crd.yaml
+    ├── 0000_10_config-operator_01_image.crd.yaml
+    ├── 0000_10_config-operator_01_infrastructure.crd.yaml
+    ├── 0000_10_config-operator_01_ingress.crd.yaml
+    ├── 0000_10_config-operator_01_network.crd.yaml
+    ├── 0000_10_config-operator_01_oauth.crd.yaml
+    ├── 0000_10_config-operator_01_project.crd.yaml
+    ├── 0000_10_config-operator_01_scheduler.crd.yaml
+    ├── 0000_10_quota-openshift_01_clusterresourcequota.crd.yaml
```
Comment (deads2k) [inline enhancements/centralized-manifest-openapi-generation.md]: no

**Pull request merged**